### PR TITLE
Added custom IAM role statements

### PR DESCRIPTION
**Pull request by eahefnawy**
Added functionality to add custom IAM role statements to the IAM role assumed by the service functions. Resolves issue #1535.

I also removed outdated `${project}` `${stage}` & `${region}` variable references that are relics from version 0. The strings get replaced anyway in code

### CODE REVIEW

**Comment by pmuens** [comment]
Love it! Such a great and nifty fix (feels like the `provider` property was exactly made for this! <3

GTM from my side!

File: lib/plugins/aws/deploy/lib/core-cf.json [hunk 1]
```diff
@@ -30,7 +30,7 @@
     "IamPolicyLambda": {
       "Type": "AWS::IAM::Policy",
       "Properties": {
-        "PolicyName": "${stage}-${project}-lambda",
+        "PolicyName": "",
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/lib/core-cf.json]: why remove policyName?

File: lib/plugins/aws/deploy/lib/core-cf.json [hunk 2]
```diff
@@ -41,7 +41,7 @@
                 "logs:CreateLogStream",
                 "logs:PutLogEvents"
               ],
-              "Resource": "arn:aws:logs:${region}:*:*"
+              "Resource": ""
```
Comment (flomotlik) [inline lib/plugins/aws/deploy/lib/core-cf.json]: Why remove limitations here?

File: lib/plugins/aws/deploy/lib/core-cf.json (hunk 1 above)
Comment (eahefnawy) [inline lib/plugins/aws/deploy/lib/core-cf.json]: it's an outdated string from v0 that gets replaced anyway in code. [Check this line](https://github.com/serverless/serverless/blob/master/lib/plugins/aws/deploy/lib/initializeResources.js#L24)

Comment (flomotlik) [inline lib/plugins/aws/deploy/lib/core-cf.json]: ok great

File: docs/guide/custom-provider-resources.md [hunk 3]
```diff
@@ -116,6 +116,28 @@ resources:
...
+  name: aws
+  iamRoleStatements:
+    - "Effect": "Allow",
+       "Action": 
+          - "something:SomethingElse"
+       "Resource": "arn:aws:logs:xxx:*:*"
+    - "Effect": "Allow",
+       "Action": 
+          - "different:VeryDifferent"
+       "Resource": "arn:aws:logs:xxx:*:*"
```
Comment (flomotlik) [inline docs/guide/custom-provider-resources.md]: Please add proper examples, e.g. 

```
  iamRoleStatements:
    -  "Effect": "Allow"
       "Action":
         - "s3:ListBucket"
       "Resource": "arn:aws:s3:::someBucket/*"
```

Comment (flomotlik) [inline docs/guide/custom-provider-resources.md]: Indentation after - needs to be two spaces btw

File: lib/plugins/aws/deploy/lib/core-cf.json (hunk 2 above)
Comment (pmuens) [inline lib/plugins/aws/deploy/lib/core-cf.json]: I think it's the same reason as above (legacy code).

File: docs/guide/custom-provider-resources.md [hunk 4]
```diff
@@ -116,6 +116,24 @@ resources:
...
+# serverless.yml
+
+service: new-service
+provider:
+  name: aws
+  iamRoleStatements:
+      -  Effect: "Allow"
+         Action:
+           - "s3:ListBucket"
+         Resource: { "Fn::Join" : ["", ["arn:aws:s3:::", { "Ref" : "ServerlessDeploymentBucket"} ] ] }
```
Comment (pmuens) [inline docs/guide/custom-provider-resources.md]: Could also be written in YAML (but might confuse the user)

**Comment by pmuens** [comment]
Reviewed it once again. GTM :shipit:

**Comment by ac360** [comment]
+1.  This looks like a nice fix and it's also what I had in mind.

**Comment by danhumphrey** [comment]
@eahefnawy outside of serverless I've been using granular (per function) IAM roles. from the docs it appears that this isn't possible - can you confirm? Am I limited to per-service roles? How are others handling granularity of permissions?

Thanks!

**Pull request merged**